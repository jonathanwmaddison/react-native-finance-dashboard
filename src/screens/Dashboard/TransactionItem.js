import React from 'react';
import moment from 'moment';
import currency from 'currency.js';
import { Card, Divider } from 'react-native-elements';
import { View, StyleSheet, Text } from 'react-native';

const styles = StyleSheet.create({
  date: {
    paddingBottom: 10,
  },
  divider: {
    marginBottom: 10,
  },
  transactionDetails: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  detailsCard: {
    width: 200,
  },
  text: {
    flex: 1,
  },
});

const TransactionItem = ({
  transTime,
  transAmt,
  description,
  transTo,
  transFrom,
}) => (
  <Card style={{ flex: 1 }}>
    <Text style={styles.date}> {moment(transTime).format('MMMM Do')}</Text>
    <Divider style={styles.divider} />
    <View style={styles.transactionDetails}>
      <View style={styles.detailsCard}>
        <Text style={styles.text}>From {transFrom}</Text>
        <Text style={styles.text}>To {transTo} </Text>
      </View>
      <View style={styles.detailsCard}>
        <Text style={styles.text}>{description}</Text>
        <Text style={styles.text}>
          {transAmt > 0
            ? '+' + currency(transAmt).format(true)
            : '-' + currency(transAmt * -1).format(true)}
        </Text>
      </View>
    </View>
  </Card>
);

export default TransactionItem;
