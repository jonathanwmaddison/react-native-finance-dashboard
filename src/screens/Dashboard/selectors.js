import { createSelector } from 'reselect';
import _ from 'lodash';
import Fuse from 'fuse.js';

const searchableCategories = [
  'transAmt',
  'description',
  'transTo',
  'transFrom',
];
const searchOption = {
  keys: searchableCategories,
  threshold: 0.3,
  location: 0,
  tokenize: true,
  maxPatternLength: 32,
  minMatchCharLength: 1,
};
// selector to help filter and sort transaction data
// fuse uses fuzzy search to find matches among searchableCategories
export const getTransactions = (state, props) => {
  const { sortBy, search } = props;

  const {
    dashboard: { transactions },
  } = state;

  if (search) {
    const fuse = new Fuse(transactions, searchOption);
    const searchResults= fuse.search(search);
    return _.orderBy(searchResults, [sortBy], ['asc']);
  } else {
    return _.orderBy(transactions, [sortBy], ['asc']);;
  }
};

export const getTransactionsState = createSelector(
  getTransactions,
  transactions => transactions
);
