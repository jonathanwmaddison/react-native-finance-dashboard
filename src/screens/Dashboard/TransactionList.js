import React, { Component } from 'react';
import { connect } from 'react-redux';
import { shape, number, string, arrayOf, bool } from 'prop-types';
import { View, FlatList, Text, ActivityIndicator } from 'react-native';
import TransactionItem from './TransactionItem';
import { getTransactionsState } from './selectors';

export class TransactionList extends Component {
  static propTypes = {
    transactions: arrayOf(
      shape({
        transId: number.isRequired,
        custId: number.isRequired,
        transTime: number.isRequired,
        description: string.isRequired,
        transTo: string.isRequired,
        transFrom: string.isRequired,
      }).isRequired
    ),
    ascendingSort: bool.isRequired,
    sortByKey: string.isRequired,
    search: string
  };

  static defaultProps = {
    search: null
  }

  keyExtractor = (item) => item.transId.toString();

  render() {
    const { transactions, isFetching } = this.props;
    if (isFetching) {
      return <ActivityIndicator />
    }
    return (
      <View style={{ flex: 1}}>
        <FlatList
          data={transactions}
          keyExtractor={this.keyExtractor}
          renderItem={({ item }) => (
            <TransactionItem
              {...item}
            />
          )}
        />
        </View>
    );
  }
}
const mapStateToProps = (state, props) => {
  return {
    transactions: getTransactionsState(state, props),
    isFetching: state.dashboard.isFetching
  }
};

export default connect(mapStateToProps)(TransactionList);
