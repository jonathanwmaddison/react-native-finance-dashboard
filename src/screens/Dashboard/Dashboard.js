import React from 'react';
import { View, Text } from 'react-native';
import _ from 'lodash';
import { SearchBar, ButtonGroup } from 'react-native-elements';
import TransactionList from './TransactionList';

const sortKeys = {
  0: 'transAmt',
  1: 'transTime',
};

const Dashboard = ({ handleSortPress, handleSearchInput, sortBy, search }) => (
  <View style={{ flex: 1}}>
    <SearchBar
      onChangeText={handleSearchInput}
      placeholder="Search your transactions..."
      lightTheme
    />
    <ButtonGroup
      onPress={handleSortPress}
      selectedIndex={sortBy}
      buttons={['Sort Amount', 'Sort Date']}
      containerStyle={{ height: 30, marginTop: 20 }}
      selectedTextStyle={{ fontWeight: 'bold' }}
    />
    <TransactionList sortBy={sortKeys[sortBy]} search={search} />
  </View>
);

export default Dashboard;
