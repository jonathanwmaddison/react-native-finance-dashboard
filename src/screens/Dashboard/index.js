import React, { Component } from 'react';
import { func } from 'prop-types';
import { View } from 'react-native';
import { connect } from 'react-redux';
import { Header } from 'react-native-elements';
import { actionCreators } from '../../services/dashboard/duck';
import Dashboard from './Dashboard';

class DashboardContainer extends Component {
  static propTypes = {
    getTransactions: func.isRequired,
  }

  componentDidMount() {
    this.props.getTransactions();
  }

  state = {
    sortBy: 1,
    search: null,
  };

  handleSortPress = index => {
    this.setState({
      sortBy: index,
    });
  };

  handleSearchInput = search => {
    this.setState({
      search,
    });
  };

  render() {
    const { sortBy, search } = this.state;
    return (
      <Dashboard
        handleSortPress={this.handleSortPress}
        handleSearchInput={this.handleSearchInput}
        sortBy={sortBy}
        search={search}
      />
    );
  }
}

const mapDispatchToProps = {
  getTransactions: actionCreators.getTransactionHistory,
};

export default connect(
  null,
  mapDispatchToProps
)(DashboardContainer);
