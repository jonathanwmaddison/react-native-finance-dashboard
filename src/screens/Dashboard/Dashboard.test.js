import React from 'react';
import renderer from 'react-test-renderer'
import { sortButtons } from './Dashboard';
import { FontAwesome } from '@expo/vector-icons';

jest.mock('@expo/vector-icons', () => ({
  FontAwesome: 'FontAwesome'
}));

describe('sort buttons', () => {
  it('renders properly when ascending, sortBy Amount', () => {
    const ascending = true
    const buttons = sortButtons(true, 0);
    const AmountButton = buttons[0].element
    expect(renderer.create(<AmountButton />).toJSON()).toMatchSnapshot()
  });
  it('renders properly when descending, sortBy Amount', () => {
    const ascending = true
    const buttons = sortButtons(false, 0);
    const AmountButton = buttons[0].element
    expect(renderer.create(<AmountButton />).toJSON()).toMatchSnapshot()
  });
  it('renders properly when ascending, sortBy Date', () => {
    const ascending = true
    const buttons = sortButtons(true, 1);
    const DateButton = buttons[1].element
    expect(renderer.create(<DateButton />).toJSON()).toMatchSnapshot()
  });
  it('renders properly when descending, sortBy Date', () => {
    const ascending = true
    const buttons = sortButtons(false, 1);
    const DateButton = buttons[1].element
    expect(renderer.create(<DateButton />).toJSON()).toMatchSnapshot()
  });
})
