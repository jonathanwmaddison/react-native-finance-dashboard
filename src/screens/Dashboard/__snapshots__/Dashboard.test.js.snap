// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sort buttons renders properly when ascending, sortBy Amount 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
   
  Amount 
  <FontAwesome
    name="sort-up"
    size={20}
  />
</Text>
`;

exports[`sort buttons renders properly when ascending, sortBy Date 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
   
  Date 
  <FontAwesome
    name="sort-up"
    size={20}
  />
</Text>
`;

exports[`sort buttons renders properly when descending, sortBy Amount 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
   
  Amount 
  <FontAwesome
    name="sort-down"
    size={20}
  />
</Text>
`;

exports[`sort buttons renders properly when descending, sortBy Date 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
   
  Date 
  <FontAwesome
    name="sort-down"
    size={20}
  />
</Text>
`;
